src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -699,9 +697,13 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
               tp
           }
 
-          val classSym = classNameToSymbol(subName(c => c == ';' || c == '<'))
-          assert(!classSym.isOverloaded, classSym.alternatives)
-          val classTpe = if (classSym eq ObjectClass) ObjectTpeJava else classSym.tpe_*
+          val classTpe =
+            classNameToSymbol(subName(c => c == ';' || c == '<')) match {
+              case _: ObjectClass.type => ObjectTpeJava
+              case classSym =>
+                assert(!classSym.isOverloaded, classSym.alternatives)
+                classSym.tpe_*
+            }
```

Comment:
oh, clever. The encapsulation permits the ident `classSym` to be used later (in a pattern). However, the idiom `case _: ObjectClass.type` looks strange FSR. I would prefer "abuse of match" formulation, `case classSym if classSym eq ObjectClass` or better yet: `val classTpe = { val classSym = ...; if - else }`